visions/accessors.md

```diff
@@ -0,0 +1,348 @@
+
+Accessors can also be `consuming`.  This indicates that the containing object must end its lifetime once this access is complete.  This is useful for representing certain types of object transformations:
+
+```swift
+struct Values {
+  var dropOldest: Values {
+    consuming get {
+       let newValues = ... everything but the oldest ...
+       return newValues
+       // Current value will end it's lifetime after this returns
+    }
+  }
+}
+
+let v = Values()
+let v2 = v.dropOldest
+// `v` is no longer alive here
+```
+
+Note that `consuming` makes no sense for setters — there is no point to changing a property on a value and then immediately ending the lifetime of that value.  Similarly, `consuming` makes no sense for `borrow` or `unsafeAddress` accessors — those require that the containing value survive for the duration of the returned borrow access or address, so it does not make sense to explicitly terminate the value lifetime immediately.
+
+This only leaves `consuming get` and `consuming read` as meaningful combinations.  A `consuming get` can be used for operations such as the `dropOldest` example above that model a transformation by creating a new value and terminating the old one.  The `consuming read` variant can be used similarly.  Unlike `borrow`, the `read` operation coroutine structure forces the containing value to live for a certain period of time, the value is consumed only at the end of that coroutine.
+
+## Some Observations
+
+Based on the above considerations, we can make a few useful observations about how these different accessors complement one another:
+
+#### Diversity is needed
+
+No one pair of accessors suffices to cover the needs of the language:
+
+* `get`/`set` require the property value to be copyable.
+* `unsafeAddress`/`unsafeMutableAddress` and `borrow`/`mutate` require the value to be already available in memory.  In particular, the mutating forms cannot be used to store a new value into a collection such as `Dictionary`, since the storage for such a value would not already be initialized.
+* `read`/`modify` exposes the value only for the lifetime of a coroutine, which is itself limited by implementation restrictions.
+
+#### Some combinations are nonsensical
+
+* Providing both a `mutate` and a `modify` for the same property doesn’t make sense.  If `mutate` is possible, then you can provide the value without making a copy, which means `modify` is unnecessary and redundant.
+
+#### We need to support existing APIs
+
+The current Swift API seems to require the following:
+
+* `set` is needed to be able to initialize new values for subscript operations.  `modify` is capable of doing this for types that can be trivially initialized, but there can be considerable overhead to construct an empty placeholder, expose it for mutation, and then record the final result.
```

Comment:
What combinations are legal/intended to be legal? It sounds like you can have a property that has both `set` and `modify`? If that's the case, are there interesting guarantees about the accessor that Swift will pick for a given operation?

Reply:
You can certainly have situations where there is more than one accessor that could be used to satisfy a particular access.  I should track down someone who knows the current rules and see if I can lay them out here.